{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "6d361f4f-f30a-434d-bef1-f1cee460f740",
   "metadata": {},
   "source": [
    "# Metasmith Starter Notebook\n",
    "Welcome! This is a starter notebook provided for you to interact with Metasmith.\n",
    "If you get stuck, you can read [the docs](https://metasmith.readthedocs.io/en/latest/), or check out [the repository on GitHub](https://github.com/hallamlab/Metasmith)!"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2508bda5",
   "metadata": {},
   "source": [
    "To begin, import the tools and structures needed from the Metasmith python API. We will also load some resources from the standard library using `Std()`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "b6eb473f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MG1655\n",
      "DH10B\n",
      "EPI300\n"
     ]
    }
   ],
   "source": [
    "import re\n",
    "\n",
    "tests = \"\"\"\n",
    "Escherichia coli str. K-12 substr. MG1655, complete genome.\n",
    "Escherichia coli str. K-12 substr. DH10B, complete sequence.\n",
    "Escherichia coli strain EPI300 chromosome, complete genome.\n",
    "\"\"\"\n",
    "tests = [x for x in tests.split(\"\\n\") if x]\n",
    "for x in tests:\n",
    "    name = None\n",
    "    for g in re.finditer(r\"substr\\.?\\s?(\\w+)|strain\\s?(\\w+)\", x):\n",
    "        a, b = g.group(1), g.group(2)\n",
    "        name = a if a else b\n",
    "        break\n",
    "    if not name: name = x\n",
    "    print(name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ddd8ded-5498-4052-8922-f0c9e26fbe8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "from metasmith.python_api import Agent, Source, Std, DataInstanceLibrary\n",
    "from metasmith.python_api import Resources, Size, Duration\n",
    "\n",
    "dtypes, containers, transforms = Std()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e1815fc2",
   "metadata": {},
   "source": [
    "Metasmith executes workflows through agents on your behalf. Each agent is given a workspace which may be on a remote machine. Let's make one called \"smith\"."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8aa78060-3311-4342-8a96-8defe0620310",
   "metadata": {},
   "outputs": [],
   "source": [
    "smith = Agent(\n",
    "    home = Source.FromLocal(Path(\"./local_home\").resolve()),\n",
    ")\n",
    "# smith.Deploy()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "847b9808",
   "metadata": {},
   "source": [
    "For this demo, we will use long reads from the model organism *Eschichia coli* EPI300, but we only have its SRA accession \"SRR35110061\". For now, all inputs must be files so let's create one with the EPI300 accession number."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ed5f734-2d45-444b-846b-bdaedd9ee3fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputs_folder = Path(\"./std_assembly_data\")\n",
    "inputs_accession_file = Path(\"./epi300.acc\")\n",
    "with open(inputs_accession_file, \"w\") as f:\n",
    "    f.write(\"SRR35110061\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6767cc26",
   "metadata": {},
   "source": [
    "We need to register the input into Metasmith's ecosystem by givging it a datatype. This lists all data types with \"accession\" in its name."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "99712dca",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in dtypes.types:\n",
    "    if \"accession\" not in k: continue\n",
    "    print(k)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a611c3dc",
   "metadata": {},
   "source": [
    "Using the `long_reads_accession` datatype, create a `DataInstanceLibrary`. This structure keeps track of multiple files and is essentially a filesystem folder managed by Metasmith."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "57cec184",
   "metadata": {},
   "outputs": [],
   "source": [
    "inputs = DataInstanceLibrary(inputs_folder)\n",
    "inputs.AddItem(inputs_accession_file.resolve(), \"std::long_reads_accession\")\n",
    "inputs.Save()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9824bdb1",
   "metadata": {},
   "source": [
    "Let's see what annotations are available. We will use \"busco_annotations\" since it will be the fastest to process."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f72a5b46",
   "metadata": {},
   "outputs": [],
   "source": [
    "for k in dtypes.types:\n",
    "    if \"annotations\" not in k: continue\n",
    "    print(k)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba9e1689",
   "metadata": {},
   "source": [
    "We can now ask the metasmith agent to generate a workflow that produces \"busco_annotations\" from a \"long_reads_accession\" using available resources and transform steps. The generated workflow with references to requried inputs are stored in `task`.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "f0ee5ef5-81bf-4037-8e88-7678dc769747",
   "metadata": {},
   "outputs": [],
   "source": [
    "task = smith.GenerateWorkflow(\n",
    "    samples=[inputs],\n",
    "    resources=[containers],\n",
    "    transforms=[transforms],\n",
    "    targets=[\n",
    "        dtypes[\"busco_annotations\"],\n",
    "    ],\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e6acc56",
   "metadata": {},
   "source": [
    "The generated plan can be viewed with graphviz. We should inspect it and ensure it is sensible."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50ffad7a-9944-4f90-be8c-2ad18ae11807",
   "metadata": {},
   "outputs": [],
   "source": [
    "from IPython.display import Image\n",
    "dagf = Path(\"dag\")\n",
    "task.plans[0][0].RenderDAG(dagf, format=\"png\")\n",
    "Image(filename=f\"{dagf}.png\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f3b8a2a",
   "metadata": {},
   "source": [
    "Staging the task transfers required files over to the agent's workspace."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d103da4b",
   "metadata": {},
   "outputs": [],
   "source": [
    "smith.StageWorkflow(task, on_exist=\"clear\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54e6d8a8",
   "metadata": {},
   "source": [
    "Before starting the run, let's alter some resource constraints and specify the local executor."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "11583b36",
   "metadata": {},
   "outputs": [],
   "source": [
    "for path, tr in transforms.IterateTransforms():\n",
    "    print(tr.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "907831a1",
   "metadata": {},
   "outputs": [],
   "source": [
    "smith.RunWorkflow(\n",
    "    task,\n",
    "    config_file=smith.GetNxfConfigPresets()[\"local\"],\n",
    "    resource_overrides={\n",
    "        \"all\": Resources(\n",
    "            cpus=8,\n",
    "        ),\n",
    "        transforms[\"busco_ref\"]: Resources(\n",
    "            cpus=2,\n",
    "        ),\n",
    "        transforms[\"fasterq_long\"]: Resources(\n",
    "            cpus=2,\n",
    "        ),\n",
    "    }\n",
    ")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "763c3775",
   "metadata": {},
   "source": [
    "The task will execute asynchronously and its progress can be checked with:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ed38a7c",
   "metadata": {},
   "outputs": [],
   "source": [
    "smith.CheckWorkflow(task)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "msm",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
